keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+<!-- /toc -->
+
+## Summary
+
+Consistent reads may be served from cache so long as:
+- A consistent (quorum) read is first made to etcd to get the latest "revsion"
+- The data in the watch cache no older than the latest "revsion" just from etcd
+
+etcd 3.4 contains a [progress notify](https://github.com/etcd-io/etcd/issues/9855)
+feature that was added specifically to make it easy to check if a cache that is
+updated via an etcd watch is up-to-date relative to some etcd revsion.
+
+This KEP summarizes how we can take advantage of progress notify feature in etcd
+3.4 to efficiently determine how up-to-date kubernetes watch caches are then
+serve reads from the watch cache when they are sufficiently up-to-date.
+
+## Motivation
+
+Intuitively, we expect that serving reads from the watch cache to be more
+performant and scalable than reading them from etcd, deserializing them,
+converting them to the desired type and then garbage collecting all the
+objects that were allocated during the read. 
+
+We will need to measure the impact to performance and scalability, but we
+believe there is a significant opportunity and would like to explore it.
+
+In addition to the general scale and performance opportunity, we also aim
+to resolve a specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) is still open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted, which can
+curently happen if the initial list request is served from a stale watch cache
+with data much older than the reflector has previously observed or if the
+api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Serve consistent reads from the watch cache for etcd 3.4 or newer
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scailability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+- Serve consistent reads from the watch cache for etcd 3.3 or earlier
+
+## Proposal
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
+
+Since etcd progress notify is only available in etcd 3.4. We would only serve consistent reads from cache if `--storage-backend` is set to `etcd3.4` or higher.
```

Comment:
Re my comment above - that would mean we would have to parse this flag to detect it - won't introducing a dedicated bool flag be easier for it?

And more explicit?

Reply:
We'll need a feature gate for progress notify. I've clarified that in the doc a bit better. For alpha, the `WatchCacheConsistentReads` flag will be off by default. And once in beta, it would be on by default but still possible to disable by default.

And once the feature is beta, it's would only enabled if the the detected etcd version is new enough. I've rewritten the section about etcd version detection to match.